classes/form/edit_entry.php

```diff
@@ -81,4 +98,30 @@ public function definition() {
         $this->add_action_buttons();
     }
 
+    /**
+     * Form validation.
+     *
+     * @param array $data
+     * @param array $files
+     * @return array
+     * @throws \coding_exception
+     */
+    public function validation($data, $files) {
+        $errors = [];
+
+        // If there is no period, then add an error to the array.
+        $periodid = period::get_period_for_date($data['completiondate'], $this->cpdlogbookid);
+        if ($periodid == 0) {
+            $errors['completiondate'] = get_string('noperiods', 'mod_cpdlogbook');
```

Comment:
Until we later add a capability around editing past dates, these two errors are effectively the same to the end user. I think it would be better if the error message say exactly 'your date needs to be between A and B' so they know how to fix it